src/test/java/com/google/cloud/teleport/util/TokenNestedValueProviderTest.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.teleport.util;
+
+import static junit.framework.Assert.assertEquals;
+
+import com.google.cloud.teleport.templates.common.SplunkTokenSource;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Tests for {@link TokenNestedValueProvider}. */
+@RunWith(JUnit4.class)
+public class TokenNestedValueProviderTest {
+
+  @Rule public ExpectedException expectedException = ExpectedException.none();
+
+  /**
+   * Test that {@link TokenNestedValueProvider#resolveTokenSource()} correctly resolves the token
+   * source as {@code SplunkTokenSource.KMS} when a token and KMS Key are provided.
+   */
+  @Test
+  public void testResolveTokenSource_kms() {
+    TokenNestedValueProvider token =
+        new TokenNestedValueProvider(
+            ValueProvider.StaticValueProvider.of(null),
+            ValueProvider.StaticValueProvider.of(
+                "projects/test-project/locations/test-region/keyRings/test-key-ring/cryptoKeys/test-key"),
+            ValueProvider.StaticValueProvider.of("test-token"),
+            ValueProvider.StaticValueProvider.of(null));
+
+    SplunkTokenSource tokenSource = token.resolveTokenSource();
+
+    assertEquals(SplunkTokenSource.KMS, tokenSource);
+  }
+
+  /**
+   * Test that {@link TokenNestedValueProvider#resolveTokenSource()} correctly resolves the token
+   * source as {@code SplunkTokenSource.PLAINTEXT} when only a plaintext token is provided.
+   */
+  @Test
+  public void testResolveTokenSource_plaintext() {
+    TokenNestedValueProvider token =
+        new TokenNestedValueProvider(
+            ValueProvider.StaticValueProvider.of(null),
+            ValueProvider.StaticValueProvider.of(null),
+            ValueProvider.StaticValueProvider.of("test-token"),
+            ValueProvider.StaticValueProvider.of(null));
+
+    SplunkTokenSource tokenSource = token.resolveTokenSource();
+
+    assertEquals(SplunkTokenSource.PLAINTEXT, tokenSource);
+  }
+
+  /**
+   * Test that {@link TokenNestedValueProvider#resolveTokenSource()} fails when a Secret Manager
+   * Secret ID is passed without a token source.
+   */
+  @Test
+  public void testResolveTokenSource_secretManager() {
```

Comment:
Can you please add one more positive test case for SECRET_MANAGER, i.e. when the `tokenSource`and `tokenSecretId` are both provided, it shouldn't fail and should return `SplunkTokenSource.SECRET_MANAGER`.

Otherwise all looks good.

Reply:
Couldn't add a test case for the exact scenario you mentioned since we decided `resolveTokenSource()` will only resolve a missing source for KMS and PLAINTEXT in [this comment](https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/274#discussion_r677814129), but I added another test case that checks whether an instance of `SecretManagerValueProvider` is returned if both `tokenSource` and `tokenSecretId` are provided.

Also added additional similar test cases for the KMS and PLAINTEXT scenarios as well.